Userland/Libraries/LibAudio/ConnectionFromClient.cpp

```diff
@@ -1,48 +1,120 @@
-// Real-time audio may be improved with a lower value.
-static timespec g_enqueue_wait_time { 0, 10'000'000 };
-
 ConnectionFromClient::ConnectionFromClient(NonnullOwnPtr<Core::Stream::LocalSocket> socket)
     : IPC::ConnectionToServer<AudioClientEndpoint, AudioServerEndpoint>(*this, move(socket))
+    , m_buffer(make<AudioQueue>(MUST(AudioQueue::try_create())))
+    , m_user_queue(make<UserSampleQueue>())
+    , m_background_audio_enqueuer(Threading::Thread::construct([this]() {
+        // All the background thread does is run an event loop
+        Core::EventLoop enqueuer_loop;
+        m_enqueuer_loop = &enqueuer_loop;
+        enqueuer_loop.exec();
+        m_enqueuer_loop = nullptr;
+        return (intptr_t) nullptr;
+    }))
+{
+    m_background_audio_enqueuer->start();
+    set_buffer(*m_buffer);
+}
+
+ConnectionFromClient::~ConnectionFromClient()
+{
+    die();
+}
+
+void ConnectionFromClient::die()
+{
+    // We're sometimes getting here after the other thread has already exited and its event loop does no longer exist.
+    if (m_enqueuer_loop != nullptr) {
+        m_enqueuer_loop->wake();
+        m_enqueuer_loop->quit(0);
+    }
+    (void)m_background_audio_enqueuer->join();
+}
+
+ErrorOr<void> ConnectionFromClient::async_enqueue(FixedArray<Sample>&& samples)
 {
+    update_good_sleep_time();
+    m_user_queue->append(move(samples));
+    // Wake the background thread to make sure it starts enqueuing audio.
+    if (!m_audio_enqueuer_active.load())
+        m_enqueuer_loop->post_event(*this, make<Core::CustomEvent>(0), Core::EventLoop::ShouldWake::Yes);
+    async_start_playback();
+
+    return {};
+}
+
+void ConnectionFromClient::update_good_sleep_time()
+{
+    auto sample_rate = static_cast<double>(get_sample_rate());
+    auto buffer_play_time_ns = 1'000'000'000.0 / (sample_rate / static_cast<double>(AUDIO_BUFFER_SIZE));
+    // A factor of 1 should be good for now.
+    m_good_sleep_time = Time::from_nanoseconds(static_cast<unsigned>(buffer_play_time_ns)).to_timespec();
 }
 
-void ConnectionFromClient::enqueue(LegacyBuffer const& buffer)
+// Non-realtime audio writing loop
+void ConnectionFromClient::custom_event(Core::CustomEvent&)
 {
-    for (;;) {
-        auto success = enqueue_buffer(buffer.anonymous_buffer(), buffer.id(), buffer.sample_count());
-        if (success)
+    while (true) {
+        m_audio_enqueuer_active.store(true);
+        Array<Sample, AUDIO_BUFFER_SIZE> next_chunk;
+
+        if (m_user_queue->is_empty()) {
+            dbgln("Reached end of provided audio data, going to sleep");
             break;
-        nanosleep(&g_enqueue_wait_time, nullptr);
+        }
+
+        auto available_samples = min(AUDIO_BUFFER_SIZE, m_user_queue->size());
+        for (size_t i = 0; i < available_samples; ++i) {
+            next_chunk[i] = (*m_user_queue)[i];
+        }
+
+        m_user_queue->slice(available_samples);
+
+        auto result = m_buffer->try_blocking_enqueue(next_chunk, [this]() {
```

Comment:
Because of the use of the wait_function here in try_blocking_enqueue, the comment in the commit description about yield() should be removed.

Reply:
That being said, I think a FIXME is in order to add a non-sleepy blocking mechanism. For example, a hardware FIFO often has a "fill level" interrupt it sends to a bare metal/kernel application. When the queue fills up all the way, the next time the queue drains to e.g. 2/3 full, it will send an interrupt to say "ok, try sending some more data if you have any".

For example for Reals, [this hardware FIFO](https://www.xilinx.com/support/documentation/ip_documentation/axi_fifo_mm_s/v4_1/pg080-axi-fifo-mm-s.pdf) generates a `Transmit FIFO Programmable Empty` interrupt "[...] when the difference between the read and write pointers of the transmit FIFO reaches the programmable EMPTY threshold value when the FIFO is being emptied." I'm sure realtime audio subsystems have some kind of similar notification they use for this case :)

Reply:
> That being said, I think a FIXME is in order to add a non-sleepy blocking mechanism.
+1